### PR TITLE
feat: add basic browser support 

### PR DESCRIPTION
**Pull request by GlenDC**
- use shims + `browser` field to support
- add `registerTestcaseResult` to communicate end of test with a node runner when running in-browser
- Expose `getEnvParameters` to let a client gather and move testground variables (used for browser support).

Supersede https://github.com/testground/sdk-js/pull/25, having applies all the previous feedback.

### CODE REVIEW

File: src/shims/runtime/logger.js
Comment (laurentsenta) [inline src/shims/runtime/logger.js:51]: (commenting here to get a thread)

Looks like a surgical change, thanks for using this approach 🚀

The CI is broken, here are the commands we run:
https://github.com/testground/sdk-js/blob/master/.github/workflows/main.yml

Comment (laurentsenta) [reply src/shims/runtime/logger.js:51]: (if you have the bandwidth, one follow-up PR / Task, could be to add a testground test to this repo, similar to what we do in the rust-sdk repo)

**Comment by laurentsenta** [comment]
cc @achingbrain if you want to take a look

File: src/shims/runtime/logger.js
Comment (laurentsenta) [inline src/shims/runtime/logger.js:47]: No strong opinion here, but could you return `console` since the API is identical?

Comment (GlenDC) [reply src/shims/runtime/logger.js:47]: Sure, was what I did originally, but thought I might more control and perhaps add some formatting and the like, but sure, console might just as well work I suppose.

File: src/shims/runtime/logger.js
Comment (GlenDC) [reply src/shims/runtime/logger.js:51]: Created an issue for it: https://github.com/testground/sdk-js/issues/27

Will do it once the browser-node plan example is in a "finished" state or whenever I find extra bandwidth sooner.

**Comment by GlenDC** [comment]
@laurentsenta this one is ready for a re-review! All checks pass now and comments have been applied.

**Review by GlenDC — COMMENTED** [review]
LGTM

**Comment by GlenDC** [comment]
Added functionality to be able to move args from host Env to Browser Env, without having to expose the required keys explicitly: https://github.com/testground/sdk-js/pull/26/commits/06df3ce6c9c904d52fa6d081aa1ffd204b5de69a

(required for: https://github.com/testground/testground/pull/1502#discussion_r1010661571)

**Comment by GlenDC** [comment]
You can see this version of `@testground/sdk` active and working in an actual node-browser cross/dual testground test plan: https://github.com/testground/testground/pull/1502

File: src/runtime/events.js
Comment (laurentsenta) [inline src/runtime/events.js:86]: Why did we remove these?

File: src/shims/runtime/logger.js
Comment (laurentsenta) [inline src/shims/runtime/logger.js]: I thought we'd just return `console`, why are we using this code? It's quite complex, with optimization, stringify and custom log functions.

File: src/env/index.js
Comment (laurentsenta) [inline src/env/index.js:42]: 👍

**Review by laurentsenta — CHANGES_REQUESTED** [review]
Thanks for recreating the PR, it looks great now, two last questions before we can merge:

- [x] Could we keep the TODO comments if they were not introduced in this PR? I don't have context on these.
- [x] Could we go back to the initial logging approach? I'd rather keep it simple & use the code before implementing optimizations, json outputs, etc. That might be related to https://github.com/testground/testground/issues/1355#issue-1271793124

File: src/shims/runtime/logger.js
Comment (GlenDC) [reply src/shims/runtime/logger.js]: Because the way we actually log from our codebase, using just `console` wouldn't give you any info at all, as it doesn't handle the given arguments well. Like that you just see things like `logged IMessage (IEvent)` or w/e, without actually seeing what's inside.

I agree that it's not as simple as using `console`, but neither do I think it's "that complex". I think this is the price you pay or not adding that support directly to winston as a contribution. That or not using winston at all, but that will result in even more code by using a complete solution.

**Comment by GlenDC** [comment]
> Thanks for recreating the PR, it looks great now, two last questions before we can merge:
> 
> * [ ]  Could we keep the TODO comments if they were not introduced in this PR? I don't have context on these.
> * [ ]  Could we go back to the initial logging approach? I'd rather keep it simple & use the code before implementing optimizations, json outputs, etc. That might be related to [testground/testground#1355 (comment)](https://github.com/testground/testground/issues/1355#issue-1271793124)

1. I'll add the TODO comments back. Reason I removed them is because my linter complained about them... Also, it's like 2 years old, IMHO if a TODO is that long in a codebase it really isn't that important and you can get to it i one day you need it, but then it will present itself. I'm just a guest in this codebase however, so I'll simply add them back :)
2. I would prefer not. The initial logger approach didn't give any useful logs at all. It's equivalent to just logging `'log'` as you get no info at all, due to how testground-sdk is using the logger...

File: src/runtime/events.js
Comment (GlenDC) [reply src/runtime/events.js:86]: Linter was complaining about them. But okay, I'll re-add them back.

File: src/shims/runtime/logger.js
Comment (laurentsenta) [reply src/shims/runtime/logger.js]: Thanks for the details, I understand why we need to move around JSON objects.

I still find the code is more complex than it should be, we're already optimizing 4 different cases, with 6 different branches. If you think that's really necessary, could you add a test for each of these cases?

it looks like we need at least:

```
logger.warn()
logger.warn('helloworld')
logger.warn({some: 'object', with: 'values'})
logger.warn('message', 'additional', 'parameters')
logger.warn('message', {additional: 'parameters'))
```

File: src/runtime/events.js
Comment (laurentsenta) [reply src/runtime/events.js:86]: > I'll add the TODO comments back. Reason I removed them is because my linter complained about them... Also, it's like 2 years old, IMHO if a TODO is that long in a codebase it really isn't that important and you can get to it i one day you need it, but then it will present itself. I'm just a guest in this codebase however, so I'll simply add them back :)

100% agree,

In that case, we're taking over a library for which original authors and maintainers are not available. I'd rather be conservative for our first few PRs, until we're familiar with it:

- It's hard to tell what is critical and not critical for now.
- the WHOLE codebase hasn't been touched much for the past two years, so on average, every TODOs will be two years old, no matter how critical.

I linked your commit to https://github.com/testground/sdk-js/issues/6 if you want to keep it, at least we know no information was lost.

File: src/shims/runtime/logger.js
Comment (GlenDC) [reply src/shims/runtime/logger.js]: Resolved in 36fca7f6cac6e5407c07bd73059f51c566dc3e1b.

File: src/runtime/events.js
Comment (GlenDC) [reply src/runtime/events.js:86]: Added back.

**Review by GlenDC — COMMENTED** [review]
LGTM

**Review by laurentsenta — APPROVED** [review]

**Comment by laurentsenta** [comment]
(updated description and title and merging, thanks for the PR!)

**Comment by mxinden** [comment]
Congratulations @GlenDC and @laurentsenta for this milestone!

**Comment by BigLep** [comment]
Agreed - way to go!

**Pull request merged**
